Model/Sync/Catalog/Processor.php

```diff
@@ -698,21 +698,31 @@ protected function existDataRequest($url, $requestIds, $sqlData, $type = 'produc
 
     /**
      * Prepare products to sync their category data
-     * @param array<mixed> $data
-     * @param array<mixed> $collectionItems
+     * @param array<mixed> $productsSyncAttempt
+     * @param array<mixed> $collectionItemsToSync
      * @param array<mixed> $dataForCategorySync
      * @return array<mixed>
      */
-    protected function getProductsForCategorySync($data, $collectionItems, array $dataForCategorySync): array
+    protected function getProductsForCategorySync($productsSyncAttempt, $collectionItemsToSync, array $dataForCategorySync): array
     {
-        foreach ($data as $dataItem) {
-            foreach ($collectionItems as $item) {
-                if ($item->getId() == $dataItem['product_id']) {
-                    $dataForCategorySync[$dataItem['yotpo_id']] = $item;
-                    break;
-                }
+        $storeId = $this->coreConfig->getStoreId();
+        $retryItems = [];
+        if ($this->retryItems && isset($this->retryItems[$storeId])) {
+            $retryItems = $this->retryItems[$storeId];
+        }
+
+        $collectionItemIdToCollectionItemMap = [];
+        foreach ($collectionItemsToSync as $itemToSync) {
+            $collectionItemIdToCollectionItemMap[$itemToSync->getId()] = $itemToSync;
+        }
+
+        foreach ($productsSyncAttempt as $productSyncAttempt) {
+            $magentoProductId = $productSyncAttempt['product_id'];
+            if (!isset($retryItems[$magentoProductId])) {
```

Comment:
So essentially you are making sure that we don't return products that failed to sync to Yotpo?

- Since the function isn't "pure" and it "adds" data to the `$dataForCategorySync` variable, can we be sure that the map doesn't contain some products that are not synced to Yotpo in advance?
- What did it break in the flow? Was it a run time error? Shouldn't it be "safe" and handled so the "other" products will be synced.